docs/source/rfc/on-chain-scripting.org

```diff
@@ -0,0 +1,954 @@
+methods.
+
+Haskell has certain similarities with Rust, which would make the
+interoperability much easier.  Though not seamless, Haskell would be
+able to represent any object that Rust can represent.
+
+The only issue is that the support for the Parity Scale Codec is not
+built-in and would have to be added by the engineers.
+
+Haskell is the safest choice in this list, because a functional
+programming language can set some guarantees that no other method can.
+Haskell has a better security track record than even Rust, because
+functional programming is fundamentally concerned with side-effects
+which includes undefined behaviour, mutation safety, safety from race
+conditions, and in some situations -- logic errors[fn:4].
+
+Haskell can produce some of the smallest binaries owing to a
+sophisticated runtime that must be present, and provided it uses
+dynamic linkage.  Default static linkage even for small programs
+produces binaries comparable in size to WASM.  It can also be stored
+as source code, and compiled in place.  Owing to ease of composition
+and a terse syntax Haskell is used in code golfing.  Haskell does not
+have as many complex invariants to uphold as does Rust, thus it can do
+dynamic linkage much more simply than Rust.  However, linking
+pro-actively, which is to say =dlopen= comes [[Haskell. in
+accomplish to hard is ][with its caveats]]
+
+Transparency requires a Haskell program to be stored as source code.
+Haskell binaries are no more intelligible than any other natively
+compiled standard.  While interpreted Haskell is nowhere near as
+well-optimised as Lua, it is not necessary.
+
+Even if a source-code submission is used the compilation can be done
+asynchronously falling back onto interpretation in the cases in which
+the compilation was not done sufficiently quickly.  One might think
+that this will add complexity, but in fact this system is already in
+place to compensate for Web Assembly's design oversights: the output
+of =rustc= though unintelligible by humans is still not completely
+compiled ahead of time, and must be optimised further.
+
+One more aspect that is unique to Haskell is the ease of
+interpretability.  Of the programming languages listed so far only
+Haskell (and Scilla) are referential-transparent, meaning that the
+meaning of a statement can be replaced with the value of that
+statement and in all situations the result of the evaluation will be
+completely unchanged.  Identifying bugs in a functional codebase is
+much easier.
+
+Sand boxing is one of the key areas where Haskell shines.  All Haskell
+programs are purely functional (unlike e.g. OCaml that is functional
+by default, but retains an imperative subset).  Any deviations must be
+carried through the type system, and unlike =unsafe= in Rust, which
+can occur silently inside of functions, functions which can cause
+input and output to happen, must carry the information through the
```

Comment:
I don't think that's true, you can do crimes in Haskell too. For example, [`unsafePerformIO`](https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO) takes an `IO a` (which represents some operation that produces `a` with side-effects) and gives you `a`, with all the side-effects

I do agree however, that it is much easier to ensure that some code has no side-effects/doesn't depend on environment/uses no random

Reply:
fn::7